arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py

```diff
@@ -0,0 +1,151 @@
+from compliance.fetch import ComplianceFetcher
+
+import yaml
+
+
+class ClusterResourceFetcher(ComplianceFetcher):
+    """Fetch resources of IBM Cloud Kubernetes clusters."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Initialize the fetcher object with configuration settings."""
+        cls.config.add_evidences(
+            [
+                RawEvidence(
+                    'cluster_resources.json',
+                    'ibm_cloud',
+                    DAY,
+                    'IBM Cloud Kubernetes cluster resources'
+                )
+            ]
+        )
+        cls.resource_types = cls.config.get(
+            'org.ibm_cloud.cluster_resources.types', RESOURCE_TYPES_DEFAULT
+        )
+        cls.tempdir = tempfile.TemporaryDirectory()
+        return cls
+
+    @classmethod
+    def tearDownClass(cls):
+        """Cleanup class."""
+        cls.tempdir.cleanup()
+
+    def _get_iks_credentials(self, cluster, cluster_configs):
+        for name in cluster_configs[cluster['name']].namelist():
+            p = pathlib.Path(name)
+            if p.name.startswith('kube-config'):
+                kubeconfig = yaml.safe_load(
+                    cluster_configs[cluster['name']].read(name)
+                )
+                cluster_token = kubeconfig['users'][0]['user'][
+                    'auth-provider']['config']['id-token']
+            if p.name.endswith('.pem'):
+                cluster_configs[cluster['name']].extract(
+                    name, path=self.tempdir.name
+                )
+                ca_cert_filepath = os.path.join(self.tempdir.name, name)
+        return cluster_token, ca_cert_filepath
+
+    def _get_roks_credentials(self, cluster, api_key):
+        s = self.session(cluster['serverURL'])
+        oauth_path = '/.well-known/oauth-authorization-server'
+        resp = s.get(oauth_path)
+        resp.raise_for_status()
+        token_endpoint = resp.json()['token_endpoint']
+        oauth_server = token_endpoint.split('/')[2]
+        s = self.session(f'https://{oauth_server}')
+        token_path = (
+            '/oauth/authorize?client_id='
+            'openshift-challenging-client&response_type=token'
+        )
+        resp = s.get(
+            token_path,
+            auth=('apikey', api_key),
+            headers={'X-CSRF-Token': 'a'},
+            allow_redirects=False
+        )
+        location = resp.headers['Location']
+        keyword = 'access_token='
+        start = location.find(keyword)
+        end = location.find('&', start)
+        cluster_token = location[start + len(keyword):end]
+        s.headers.update = ({'X-CSRF-Token': None})
+
+        return cluster_token, None
+
+    @store_raw_evidence('ibm_cloud/cluster_resources.json')
+    def fetch_cluster_resource(self):
+        """Fetch cluster resources."""
+        with evidences(self.locker, 'raw/ibm_cloud/cluster_list.json') as ev:
```

Comment:
`evidences` is the context manager for checks not fetchers.  You can't do this.  I'm  not really sure how it even works.

Reply:
I don't really even understand what's happening here.  Putting aside the misuse of the `evidences` context manager.  It appears that this fetcher depends on evidence gathered by another fetcher?  That really isn't clear in the README write up.  If that is your intent then you need to use get_evidence_dependency otherwise this fetcher will not run correctly all the time.  I'm surprised that it hasn't failed for you during your testing.

Reply:
I was not clear about `with evidences`; I changed this to using [get_evidence_dependency](https://github.com/ComplianceAsCode/auditree-arboretum/pull/54/files#diff-bcd498e6b9c862fab4f20772411039245ac766ae4cab4186d5c9285b8835064aR119-R122), and the test result is as follows.

```
 $ compliance  --fetch -C ../auditree-data/cluster-list-fetcher/config.ibm_cloud.multi_account_custom_rsrc.json arboretum.ibm_cloud.fetchers.fetch_cluster_resource arboretum.ibm_cloud.fetchers.fetch_cluster_list --evidence local
INFO: Creating locker in /var/folders/s8/57ct85zd407d1g6cx0dtjhqw0000gn/T/compliance...

Fetcher Primary Run

E.
======================================================================
ERROR: fetch_cluster_resource (arboretum.ibm_cloud.fetchers.fetch_cluster_resource.ICClusterResourceFetcher)
Fetch cluster resources.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/locker.py", line 456, in get_evidence
    metadata['ttl'],
TypeError: 'NoneType' object is not subscriptable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/evidence.py", line 637, in get_evidence_dependency
    evidence = locker.get_evidence(path)
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/locker.py", line 465, in get_evidence
    f'Evidence {evidence_path} is not valid '
compliance.utils.exceptions.EvidenceNotFoundError: Evidence raw/ibm_cloud/cluster_list.json is not valid or not present within the locker

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/evidence.py", line 684, in wrapper
    return _store_wrapper(self, evidence_path, f, 'raw')
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/evidence.py", line 840, in _store_wrapper
    content = func(self)
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/arboretum/ibm_cloud/fetchers/fetch_cluster_resource.py", line 120, in fetch_cluster_resource
    'raw/ibm_cloud/cluster_list.json', self.locker
  File "/Users/tmishina/workspace/auditree-evolve/compliance-evidence-collector/venv/lib/python3.7/site-packages/compliance/evidence.py", line 659, in get_evidence_dependency
    f'evidence dependency {path} is currently unavailable.'
compliance.utils.exceptions.DependencyUnavailableError: evidence dependency raw/ibm_cloud/cluster_list.json is currently unavailable.

----------------------------------------------------------------------
Ran 2 tests in 3.030s

FAILED (errors=1)

Fetcher Dependency Re-Run #1

.
----------------------------------------------------------------------
Ran 1 test in 87.288s

OK
INFO: Committing changes to local locker in /var/folders/s8/57ct85zd407d1g6cx0dtjhqw0000gn/T/compliance...

```